### ISSUE TITLE
Adding interface that is more C++ container friendly

### ISSUE THREAD

**Issue by ajaust** [precice/precice#717]
**Please describe the problem you are trying to solve.**
preCICE is written in C++ with an C++ API that is not container friendly. However, the use of containers is often encouraged in (modern) C++. Some solver libraries/frameworks (like DUNE, for example) also make extensive use of containers.

Side note: 
I don't think we should encourage people to use raw pointers as shown in the [adapter example](https://github.com/precice/precice/wiki/Adapter-Example), but I understand that it should be part of the API for legacy codes.  I would propose to add an additional adapter example that actually uses some STL container instead of raw pointers. Maybe this should be another issue?!

**Describe the solution you propose.**

I propose function interfaces that could look like:
```c++
template< template< class ...> class CppContainer, class ... Args >
void writeBlockScalarData (
    const int dataID,
    const CppContainer<int, Args...>&    valueIndices,
    const CppContainer<double, Args...>& values );
```

This should be container friendly while ensuring that the correct datatypes (`int` and `double`) are used. The compile time of preCICE itself should be unaffected as the templated functions will be compiles with the user's code. (Note: I need the `class ... Args` for the allocator of the container) 

The containers would need to provide a `data()` member function that gives us a pointer to the (contiguously) stored data and a `size()` member that gives the size of the current member in a suitable type.

The new interface would only touch functions that accept a raw pointer at the moment. This should affect about 10 functions, if i counted correctly.

At the moment, I see three (more or less reasonable) ways to implement the interface, but I am not sure what is the best way. I would like to discuss this here:
1. Write a wrapper (could be a class) around the `SolverInterface` that has a member of `class SolverInterface` and calls it appropriately. This would be similar to the approach that is used in the C bindings for examples, see `extras/bindings/c/src/SolverInterfaceC.cpp`. 
    - Pro: Should be easy to implement. Can hide API functions I do not want to expose.
    - Con: Additional layer of function calls, but I don't think that this is a real issue.
2. Write a new class that is derived from `class SolverInterface`
    - Pro: Automatically inherit API and implementation so I don't have to reimplement functions that I do not want to use.
    - Pro/Con: Has still "C-style" interface available.
3. Write a new class in the style of `src/precice/SolverInterface.cpp` and leave out API calls that I don't want to have.
    - Pro/Con: Lots of control over the interface such it would be possible to prevent use of old API calls. 
    - Con: Has to reimplement functions that are already implemented in `src/precice/SolverInterface.cpp`.

**Describe alternatives you've considered**
I can call preCICE like this

```c++
std::vector<int> vertexIDs;
std::vector<double> data;
//stuff
assert( /* check that the cast to int is safe */ );
assert( /* Maybe check that the size of the vectors is compatible */ );
precice_->writeBlockScalarData( dataID, 
                                (int)vertexIDs_.size(), 
                                vertexIDs_.data(), 
                                data.data() );
```
but I find it a bit tedious and not very clean. Using container it would be possible simplify the arguments to the function `writeBlockScalarData` as in the example above since the container knows its own size.

**Additional context**
I am working (or at least thinking) about the proper structure of a preCICE adapter for DUNE/DuMuX. DUNE and DuMuX use a lot of containerized data structures that behave like container from the standard library, but are tailored for high computational performance. Therefore I work on some intermediate abstraction that makes it more convenient to use containers.

I wondered if you would be interested in such a feature. It should be possible to start with a drop-in class that one can add to one' project to use the container friendly interface. Afterwards one could the corresponding functions into the current preCICE interface as it does not break compatibility. One would have to discuss if that blows up the API too much. 

As I want to play with a more container friendly interface anyway, I would be happy to contribute with that in some way.

**Comment by fsimonis** [comment]
Thank you for this very detailed issue.

preCICE currently provides a C-style API, which aims to be minimal. This minimizes the documentation, simplifies the implementation of bindings and imposes the same usage patterns for adapters across languages. The latter is of course slowly fading into insignificance as we increasingly focus on usability instead of API consistency. The Matlab and Python bindings are good examples for that trend.

**General thoughts on the C++ API**

Modern C++ actually attempts to be container agnostic. The STL is a good example as the interface of its types are generally iterator-oriented (even after the C++20 ranges). Boost is more interesting in this case, as many libraries provide both an iterator and range-based interface for members as well as non-member functions.

Applying the Boost-style to the preCICE API could look like this:
```cpp
// C-Style
void writeBlockVectorData (int dataID, int size, const int *valueIndices, const double *values);

// Iterator-Style
template<class IdIter, class DataIter>
void writeBlockVectorData (int dataID, IdIter indicesBegin, IdIter indicesEnd, DataIter dataBegin, DataIter dataEnd);

// Range-Style
template<class IdRange, class DataRange>
void writeBlockVectorData (int dataID, IdRange const & indices, DataRange const & data);
```

This leads us to the following problems:
1. Contiguous Iterators were introduces in **C++17**, thus there is no safe way to simply dispatch the iterator variant to the C-Style variant. One can always define pre-conditions in the interface, but those cannot be checked.
2. Ranges provide `begin()` and `end()`, so that leads us to the previous problem.
3. Range-types providing 'data()` and `size()` could be directly dispatched to the C-Style variant. Implementing this as a special case requires to detect these functions for the passed type. This requires custom type traits and and some SFINAE in the headers and thus the public interface of preCICE.
4. Checks based on the passed types should use the internal `PRECICE_CHECK()` log macros. Thus we need to expose a lot of interna just for the sake of error-checking.

So there are a lot of problems to be solved if we aim for a complete solution.

**A Span-Only Solution**

The interface you proposed is essentially an interface accepting types following a `span`like concept. We could extend the interface by these types, but I fear that this will confuse users.

So the question is: Are we comfortable with adding a non-trivial interface for convenience?

[Implementation of the trait with example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlICsupVs1qhkAUgBMAISnTSAZ0ztkBPHUqZa6AMKpWAVwC2tLgHZSW9ABk8tTADljAI0zEQAVlIAHVAsLqtHqGJuY%2BfgF0tvZORq7uXkoqanQMBMzEBMHGppwWSZiqgWkZBNGOLm6eiumZ2aF5NaXlsfGeAJSKqAbEyBwA5FIAzHbIhlgA1OJDOgBuhUTE09jiAAwAgmvrtMxGmArezH0TeEberFNmshsEmGfCt1MzY8wKChM6y1sKBMQGqicFAB9A4iS7XdYTCYAemhEwAdOhmOkIO0JrcfgotlCfsi8MgJsgEBl0fsCAARZHMCDw2ntaYQqG3e7IzBPHQvN4TAAqpHRAE9vFpdmzpuSJlgxgRBZgID90CAQJLWLM2NMdNzlqjEVTUfShisNjj0moCXYCKSfpSUdzCXQflNJAA2fUQ7EwuHw/wAL1laIxBCxRomuNNhOJxEtjDwvppdIZ7uZ51Z7M5715AqFOz2T3FyulQrlBAVSsKKrVM01Bu1Pr9%2BsNkJDJvxJ3oUYYMdltrQtAdUhdCY27thHwQhQA1pm2ToJjRIzqUWiROgEbXUe6ez9MAAPbyR0Mt5yofQTVWGUVDcnuqHGkt4YEKEXq818/MyiAB63U%2BWK5Vn9WfNW7TAcsipngYorOv2143j%2BID3iCT4zC%2BEplgWsoBh2sZwX%2BFY6IB2B6vWYFsBBg6bGYV5DG6lFQiO2b7IcxynOcWxbCOOgALQ/Py7CzgYtBFHQWyzKgeArlQAlCbQEDmnavYWgAVEi6R8nJtaumxGwjgwhy0Nx0p8ZJgkpLQWxJg8F4csIXJ6PQzB2G4Xw3DKDFNiWwrOOwQJ4FQz73IqCGgmZMx2ekjlLAaJHnnyoniaBIDofxJmBBAMFhQ59iRpuFr9sgaLiFc7rGdJEDIAu1KdISXqdnq5GFVeQ4bHJRiZai4LunB8yqCQz70Msp6FdInB8pIfJDHyvATF4ExOnyFgTAAHA19XBhxBm8WyLCsKwxVSaZECzBVqKxTVsYgdRWmNjpekbXx227cGJUHbMmlDpR/SdKwID9B4/SkKY/SrP9qA/Z8MhyCG3S9JBQycP9BA/cDwGkOOIBOqs8KcB4TpOh4ZgAJxOkM3DcE63CLUIP3cP9gPA6QoP9P9CggKspCI0Dn2kHAsAwIgKCoGceDsGQFBlYL3jC1UwAExNVDC7cxAsxAzhI/9zh2Bk/I/fDpBoEYez0AA8rQrDa5zpBYK1ojsGrlt4MQCx4PMLMWzuhQGLcOv/eayh26weDOMQWt6Fg3vs8Qpze50ND0EwbAcDw/CCMIogoHIchCIHLOQJ0qDeKZrucQZ6BihIEMyJwusFNJmjaPUuSWNoLSVO4I2%2BP4pkN4IHeRLQLdxFUI016ZJR1PoOSCCPxS1GUjmtEPTTjyEje4pkA9tJwnQKNDfRcF9P1/QDduM9ui1Opx5MTMAyAEgT8JDBMEC4IQJCOnDfJ6ELIvv1vHwZzIBGasUZow8Gzb6/QaakCMJ4NmdMQY/WZqzdmwDuZ8wgEgboBBvCe3IJQfWksRaCEwPgRYghY6MG2onXgAB3YO3ho5U1%2BrTE%2BP1hicAmDQwgCAJhnwvlfG%2Bd8H5AM5ijMczAsDuBOhAqBMCwHHwtozJBbMObI06GjIY99FrcCGCTSQqwhh4w8EMDwY0IFDBYYoxBKDREH36JISx9MlE2LUaQeYStAggG4EAA%3D%3D%3D)

**Implementation of the Interface**

I personally have been playing a lot with the thought of creating a modern strongly-typed header-only wrapper for the preCICE library. The main problem with such a project is that it will (sooner or later) fracture the community.

We need a solution that can be integrated into the SolverInterface and coexist with the C-Style interface of the function.

**Comment by uekerman** [comment]
A few remarks:
* We should indeed port the Adapter Example (now) to use containers. `std::vector<double> forces` instead of `double* forces = new double[vertexSize*dim]`, but still using the current C++ interface. 
* I think for the moment suggestion 1) of @ajaust  sounds like a good way to start to me. Will be good to experiment as it is rather independent. We can mark it as experimental and still change things later. 
* Contributions here are very much appreciated! 
* I don't see the danger of fracture of the community. First, many people just use adapters. Second, already now everybody uses another API. My gut feeling is sth like (we did not collect any data): 
  * C++ 20%, 
  * C 10%, 
  * Fortran 35%, 
  * FortranModule 10%, 
  * Python 20%, 
  * Matlab 5%

**Comment by ajaust** [comment]
I see that it is a bit more complicated to find a good solution that is following modern C++ style in a proper way than I thought.

Maybe the solution does not have to be too elaborated either and something simpler like
```c++
template<typename SIZE_TYPE>
void writeBlockVectorData (const int dataID, const SIZE_TYPE size, const int *valueIndices, const double *values);
```
would suffice? To me it feels more tedious to use containers with preCICE than it could be in my opinion. At the moment one has to explicitly cast the return value of `size()` to `int` to avoid warnings. Before doing so one should even check whether the value it fits into `int`.

Overall, it is not a problem if the preCICE interface stays the way it is. I personally would try to hide certain casts in my adapter in that case. For the DuMuX adapter I will try this and see what works out for me. The solutions proposed by @fsimonis look actually much more like proper C++, but this might be something I would rather do exclusively in the adapter and not in the C++ interface (at the moment). 

Maybe you should just wait. If the interface becomes a topic you want/have to work on, you can get in touch with me and I will report my feedback in case I have not done this yet. :)

**Comment by fsimonis** [comment]
Also relevant for this discussion http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Ri-array

**Comment by ajaust** [comment]
I have been thinking a bit more about the issue and came up with the following observations

- I complain too much. :D 
- I completely agree with @fsimonis that my approach is not proper C++ either. It should be iterators at the moment and that there is no nice way to fix that (due the amount of data structures that have to be passed and the problems with contiguity of data). I guess I just got depressed that I have to have to always call `data()` function for containers and then cast the `size()` explicitly in order to get rid of all warnings. 
- I am always a bit afraid to cast `size_t` to `int` since the value might not fit into an `int`. I usually try to run some extra checks on this before casting. 
- Tailoring the preCICE function calls to the certain data structures of the solver should be done in the adapter. The developer of the adapter then has to deal with the proper transformation in order to pass it to the actual preCICE call. So it might be better to keep the API more or less as it is. Maybe with the adaption of the size_types and the how many size parameters the functions should expect (see the comment above).

Other things that popped up in my head:
- In functions like `writeBlockScalarData`: Why do I have to give the `vertexIDs` to preCICE the whole time. Is it a common use case that I only want the data at a subset of vertices? 
- From a user's perspective (ok... my perspective) it feels that the C++ API is in the end a C API in disguise. Besides the use of `std::string` and namespaces nothing seems to take advantage of the C++ features. Moreover, the `SolverInterface` is a class, but I still have to manually call `finalize()` myself. In this sense it is not even an object that cleans up after itself. I assume that is needed due to the order `MPI_Finalize()` must be called. In the end `SolverInterface` does not does make the use of preCICE easier from C++ easier or natural than using the C interface instead. 

> Also relevant for this discussion http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Ri-array

I support that as well!

TL;DR
- The concise API of preCICE is probably good the way it is and some of the problems should be solved in the adapters rather than in preCICE itself.
- I still complain too much.

**Comment by BenjaminRueth** [comment]
>     * Moreover, the `SolverInterface` is a class, but I still have to manually call `finalize()` myself.

Related discussion: https://github.com/precice/fenics-adapter/pull/59#issuecomment-632575122

**Comment by ajaust** [comment]
I just wanted to add that I assume the similarity of the different APIs is partially wanted in order to not have too big deviations between APIs for the different languages and possibly due to some technical reasons when writing a multi-language API. 

In this sense I guess it makes sense from the library perspective to limit the amount of features of one of the languages being used for better maintainability and usability. That is the perspective that I as a user am lacking a bit. So don't be too worried about my complaints.

**Comment by fsimonis** [comment]
@ajaust You definitely don't complain too much. Constructive criticism is very helpful!

We are always refactoring and modernising parts of preCICE, which can be quite time consuming. It is definitely good to have some hints on where to look next. I would never have expected there to be bugs like this in the bindings themselves. So cudos for finding them!
 
When I started as a developer, I had the same thought regarding the interface. It really is a C-like API to a C++ library.
This was a design decision to keep language interfaces similar.  
This consistency is difficult to maintain though. Especially the python bindings required a new interface to make them usable and not totally confusing.

Pre v2 preCICE used to have a 3 phase initialisation and a 2 phase finalisation.
( ctor, configure, initialise, and finalise, dtor )
We merged configure into the ctor in v2. Now it may be a good time to make the dtor clean up which is not an API breaking change.

The block data interface is quite general as it allows to access both subsets and chunks of data in arbitrary order. It would be very dangerous for preCICE to make assumptions on the order of vertices, especially as they cannot be checked.
You could write a helper type that keeps track of the SolverInterface and the vertex ids. Then you could just do a `helper.write(data)`.

I have been working on a little modern wrapper for the SolverInterface, but it is difficult to come up with an interface that _feels_ like C++.

**Comment by uekerman** [comment]
> @ajaust You definitely don't complain too much. Constructive criticism is very helpful!

Yes, absolutely! These discussions are good. Any easy to find for others.

> In functions like writeBlockScalarData: Why do I have to give the vertexIDs to preCICE the whole time. Is it a common use case that I only want the data at a subset of vertices?

It's not as uncommon as you might think. Think of a DG code that has an only local perspective in each element. Then, we want to read for example values for all GPs within on element. We had this exact situation when coupling Ateles.

**Comment by fsimonis** [comment]
This could be a good topic to resurrect with the upcoming preCICE 3.0 release.

If we decide to change this, then the big question is if we want to completely modernize the interface, essentially breaking the similarities between C++, C, Fortran and the like.

Possible example
```cpp
std::vector<double> coords = getCoordinates();
std::vector<double> data = getData();

auto meshA = solverinterface.mesh("meshA");
auto vertices = meshA.setVertices(coords);
vertices.write("dataOne", data);

vertices.read("dataOne", data.begin());
```

**Comment by BenjaminRodenberg** [comment]
I think we are also slowly moving towards closing this issue. What is still missing here?

**Comment by fsimonis** [comment]
I think #1641 solves the issue

**Comment by fsimonis** [comment]
After #1636 and #1641, preCICE accepts containers of contiguous memory as well as arrays directly via the interface.

```cpp
std::vector<VertexID> vids(2)
std::vector<double>   coords{0,0,0, 1,1,1};
participant.setMeshVertices("Mesh", coords, vids);

VertexID edges[] = {vids[0], vids[1]};
participant.setMeshEdges("Mesh", edges);

std::array<double, 2> data;
participant.readData("Mesh", "Scalar", dt, vids, data);
```
